Jailing, 2-step unstake, Slash and Penalty

ii-cruz opened the pull request:
## What's in this pull request?

### Jailing
After being "slashed", a validator goes to "jail" and is forced to be inactive for a certain time frame.
During this period, the validator cannot be reactivated.

### 2-step unstake
Stakers need to inactivate stake before being able to withdraw it.
Inactive stake is locked up for a certain time frame before becoming available.

Re-delegation requires the whole stake to be inactive first.

The jailing also applies to the staker.
Thus a staker can only re-delegate or unstake after the jail period and the lock-up period for inactive stake have passed.

### Slashing and Penalizing
The old slashing is renamed to penalizing. Penalties are applied to slots and deactivate the validator. The penalties are a consequence of skip blocks.
Slashing is applied to all validator slots and jails the validator. Slashes are a consequence of forks (other malicious behavior still needs to be integrated).

### Clean up of bit sets in the staking contract and macro block
Inside the staking contract, the sets are organized in a separate struct. Additionally, the macro block now only contains the necessary information for syncing.

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

ii-cruz committed:
Adding jailing to the validator

Co-authored-by: Pascal Berrang <git@paberr.net>

ii-cruz committed:
2 step unstake

Co-authored-by: Pascal Berrang <git@paberr.net>

jsdanielh commented on primitives/account/src/account/staking_contract/receipts.rs:
```diff
@@ -15,6 +15,7 @@ pub struct SlashReceipt {
     pub newly_deactivated: bool,
     pub newly_disabled: bool,
     pub newly_lost_rewards: bool,
+    pub old_jail_release: Option<u32>,
```
Could you add `rustdoc` to the members?

jsdanielh commented on primitives/account/src/account/staking_contract/receipts.rs:
```diff
@@ -27,6 +28,22 @@ pub struct UpdateValidatorReceipt {
 convert_receipt!(UpdateValidatorReceipt);
 
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct JailValidatorReceipt {
```
Could you add `rustdoc` to the `struct` and its members?

jsdanielh commented on primitives/account/src/account/staking_contract/receipts.rs:
```diff
@@ -48,6 +65,7 @@ pub struct DeleteValidatorReceipt {
     pub reward_address: Address,
     pub signal_data: Option<Blake2bHash>,
     pub inactive_since: u32,
+    pub jail_release: Option<u32>,
```
Could you add `rustdoc` to the `struct` and its members?

jsdanielh commented on primitives/account/src/account/staking_contract/receipts.rs:
```diff
@@ -56,3 +74,17 @@ pub struct StakerReceipt {
 convert_receipt!(StakerReceipt);
+
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct SetInactiveStakeReceipt {
```
Could you add `rustdoc` to the `struct` and its members?

jsdanielh commented on primitives/account/src/account/staking_contract/receipts.rs:
```diff
@@ -56,3 +74,17 @@ pub struct StakerReceipt {
+convert_receipt!(SetInactiveStakeReceipt);
+
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct RemoveStakeReceipt {
```
Could you add `rustdoc` to the `struct` and its members?

jsdanielh commented on web-client/src/transaction.rs:
```diff
@@ -663,6 +672,13 @@ pub struct PlainUpdateStakerData {
 
+#[derive(Clone, serde::Serialize, serde::Deserialize, Tsify)]
+#[serde(rename_all = "camelCase")]
+pub struct PlainSetInactiveStakeData {
```
Could you add `rustdoc` to the `struct` and its members?

ii-cruz committed:
Tests

Co-authored-by: Pascal Berrang <git@paberr.net>

paberr commented:
We will add rustdocs as part of a separate PR that will update how the sets are being used.

ii-cruz committed:
Renaming slot to slot allocation

Co-authored-by: Pascal Berrang <git@paberr.net>

ii-cruz committed:
Abstracting the lost reward and disable slots sets.

Co-authored-by: Pascal Berrang <git@paberr.net>

ii-cruz committed:
Updated the sets on the macro block body

Co-authored-by: Pascal Berrang <git@paberr.net>

paberr committed:
Create slash inherent for fork proofs

Co-authored-by: ii-cruz <ii.cruz@campus.fct.unl.pt>

paberr committed:
Don't allow to delete jailed validators
Increase waiting time for delete to always include reporting window
Generalize slash for larger reporting windows
Increase reporting window to a epoch

Co-authored-by: ii-cruz <ii.cruz@campus.fct.unl.pt>

sisou commented on web-client/src/client/account.rs:
```diff
@@ -139,11 +133,20 @@ impl PlainAccount {
+    /// Stake can only be re-delegated if the whole balance of the staker is inactive and released
+    /// (or if there was no prior delegation). For inactive balance to be released, the maximum of
+    /// the `inactive_release` and the validator's `jail_release` must have passed.
+    inactive_balance: Coin,
```
This is a `PlainStaker` object, and `Coin` does not translate cleanly to a number in JSON. Please set this as `u64` like the `balance` field.

hrxi replied:
What does `Coin` translate to in JSON? Perhaps it should translate to a simple number.

paberr replied:
Didn't #1681 change the JSON serialization of coin to just a simple number?

hrxi replied:
Yea, looks like it: https://github.com/nimiq/core-rs-albatross/pull/1681/commits/a666be6d6deb6900660be084f44960fc6d563fc8#diff-ac8e9d0ded515f48f601114227b21f3b7e850e8a850a7fdc8aad2103f8c8609e.

sisou replied:
Well, then not JSON, but to WASM. Because what I get now when compiling to WASM is these types:
```
export interface PlainStaker {
    balance: number;
    delegation: string | undefined;
    inactiveBalance: Coin;
    inactiveRelease: number | undefined;
}
```
where `Coin` is not defined and thus a type error.

sisou reviewed (changes requested):
Please add a method for setting inactive stake (e.g. `new_set_inactive_stake`) in the web-client: `web-client/src/primitives/transaction_builder.rs`.

paberr committed:
Add setInactiveStake function to web-client

Co-authored-by: ii-cruz <ii.cruz@campus.fct.unl.pt>

paberr commented:
@sisou We addressed both of your comments.

jsdanielh commented:
Rebasing it to get it merged
